quickwit/quickwit-ingest-api/src/memory_capacity.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use tokio::sync::Notify;
+
+#[derive(Clone)]
+pub struct MemoryCapacity {
+    inner: Arc<InnerMemoryCapacity>,
+}
+
+impl fmt::Debug for MemoryCapacity {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("MemoryCapacity")
+            .field("capacity", &self.capacity())
+            .field("max_capacity", &self.max_capacity())
+            .finish()
+    }
+}
+
+impl MemoryCapacity {
+    pub fn new(max_capacity: usize) -> Self {
```

Comment:
YOu have a hidden contract here.

Ideally make it 
```suggestion
    pub fn new(max_capacity: NonZeroU64) -> Self {
```
but sometimes NonZeroU64 are annoying. An alternative is to put an assert and document the contract.
or